src/spaces/grid.jl

```diff
@@ -0,0 +1,241 @@
+# the file github.com/JuliaDynamics/TimeseriesPrediction.jl/src/spatiotemporalembedding.jl
+# It creates a performant envinroment where the cartesian indices of a given neighborhood
+# of given radious and metric type are stored and re-used for each search.
+
+Base.length(r::Region{D}) where {D} = prod(r.maxi .- r.mini .+ 1)
+function Base.in(idx, r::Region{D}) where {D}
+    @inbounds for φ in 1:D
+        r.mini[φ] <= idx[φ] <= r.maxi[φ] || return false
+    end
+    return true
+end
+
+# This function calculates how large the inner region should be based on the neighborhood
+# βs and the actual array size `d`
+function inner_region(βs::Vector{CartesianIndex{D}}, d::NTuple{D,Int}) where {D}
+    mini = Int[]
+    maxi = Int[]
+    for φ in 1:D
+        js = map(β -> β[φ], βs) # jth entries
+        mi, ma = extrema(js)
+        push!(mini, 1 - min(mi, 0))
+        push!(maxi, d[φ] - max(ma, 0))
+    end
+    return Region{D}((mini...,), (maxi...,))
+end
+
+# This function initializes the standard cartesian indices that needs to be added to some
+# index `α` to obtain its neighborhood
+import LinearAlgebra
+function initialize_neighborhood!(space::GridSpace{D}, r::Real) where {D}
+    d = size(space.s)
+    r0 = floor(Int, r)
+    if space.metric == :euclidean
+        # hypercube of indices
+        hypercube = CartesianIndices((repeat([(-r0):r0], D)...,))
+        # select subset of hc which is in Hypersphere
+        βs = [β for β ∈ hypercube if LinearAlgebra.norm(β.I) ≤ r]
+    elseif space.metric == :chebyshev
+        βs = vec([CartesianIndex(a) for a in Iterators.product([(-r0):r0 for φ in 1:D]...)])
+    else
+        error("Unknown metric type")
+    end
+    inner = inner_region(βs, d)
+    whole = Region(map(one, d), d)
+    hood = Hood{D}(inner, whole, βs)
+    space.hoods[float(r)] = hood
+    return hood
+end
+
+
+"""
+    grid_space_neighborhood(α::CartesianIndex, space::GridSpace, r::Real)
+
+Return an iterator over all nodes within distance `r` from `α` according to the `space`.
+
+The only reason this function is not equivalent with `node_neighbors` is because
+`node_neighbors` skips the current position `α`, while `α` is always included in the
+returned iterator (because the `0` index is always included in `βs`).
+
+This function re-uses the source code of TimeseriesPrediction.jl, along with the
+helper struct `Hood` and generates neighboring cartesian indices on the fly,
+reducing the amount of computations necessary (i.e. we don't "find" new indices,
+we only add a pre-determined amount of indices to `α`).
+"""
+function grid_space_neighborhood(α::CartesianIndex, space::GridSpace, r::Real)
+    hood = if haskey(space.hoods, r)
+        space.hoods[r]
+    else
+        initialize_neighborhood!(space, r)
+    end
+    _grid_space_neighborhood(α, space, hood)
+end
+
+function _grid_space_neighborhood(
+    α::CartesianIndex,
+    space::GridSpace{D,true},
```

Comment:
What is this constructor `GridSpace{D,true}` vs the `GridSpace{D,false}` below? (I'm wondering if this function is ever called & perhaps should have been dropped with the rest of the TimeseriesPrediction code we aren't using)

Reply:
It is correct there. the `true` means periodic. THere is one more below with `false` for non-periodic.